### ISSUE TITLE
types.(Async)GeneratorType is incompatible as a return for (Async)Generators

### ISSUE THREAD

**Issue by Gobot1234** [microsoft/pyright#6122]
**Describe the bug**
GeneratorType is the true type of generators so it should be assignable as the return type of the function.

**Code or Screenshots**
```python
from types import AsyncGeneratorType, GeneratorType


def yields() -> GeneratorType[None, None, None]:
    yield


async def async_yields() -> AsyncGeneratorType[None, None]:
    yield
```
```
  /Users/James/Programming/steam.py/test3.py:5:5 - error: Return type of generator function must be compatible with "Generator[None, Any, Any]"
    "Generator[None, Unknown, Unknown]" is incompatible with "GeneratorType[None, None, None]" (reportGeneralTypeIssues)
  /Users/James/Programming/steam.py/test3.py:4:17 - error: Return type of generator function must be compatible with "Generator[Any, Any, Any]"
    "Generator[Any, Any, Any]" is incompatible with "GeneratorType[None, None, None]" (reportGeneralTypeIssues)
  /Users/James/Programming/steam.py/test3.py:9:5 - error: Return type of async generator function must be compatible with "AsyncGenerator[None, Any]"
    "AsyncGenerator[None, Unknown, Unknown]" is incompatible with "AsyncGeneratorType[None, None]" (reportGeneralTypeIssues)
  /Users/James/Programming/steam.py/test3.py:8:29 - error: Return type of async generator function must be compatible with "AsyncGenerator[Any, Any]"
    "AsyncGenerator[Any, Any, Any]" is incompatible with "AsyncGeneratorType[None, None]" (reportGeneralTypeIssues)
```

**VS Code extension or command-line**
pylance: v2023.10.10

**Comment by erictraut** [comment]
Pyright is working correctly here.

The types defined in the `types` module are internal-only types. If you're using these in type annotations, you're likely doing something wrong. You should use the forms defined in `typing` for annotations. Type checkers contains significant special casing for classes defined in `typing` but have no knowledge of the types in the `types` module.

Not surprisingly, mypy also generates the same errors as pyright in this case.

**Comment by Gobot1234** [comment]
This is related to the discussion in https://github.com/python/typing/issues/1480. The typing forms aren't (hopefully) going to be as useful as they aren't concrete types (https://github.com/python/typeshed/pull/10816). My intention is also to submit a PR to mypy that hopefully is able to add support for having the same behaviour as what I would hope pyright can have.

**Comment by erictraut** [comment]
Please do not submit a PR to mypy unless/until you first achieve some level of consensus within the typing community on this point. I personally think it's a bad idea.
